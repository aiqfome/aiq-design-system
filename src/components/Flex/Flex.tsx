import React from 'react'
import styled, { css, DefaultTheme } from 'styled-components'
import PropTypes from 'prop-types'
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  layout,
  LayoutProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  border,
  BorderProps,
  position,
  FlexProps,
  flex,
  PositionProps,
  flexbox,
  FlexboxProps
} from 'styled-system'

export interface Props
  extends SpaceProps,
    DefaultTheme,
    LayoutProps,
    FontSizeProps,
    FontWeightProps,
    PositionProps,
    ColorProps,
    FlexProps,
    FlexboxProps,
    BorderProps {
  variant?: 'auto' | 'centralized' | 'fullCentralized'
  color?: string
  fullHeight?: boolean
  children?: any
  className?: string
  onClick?: (e: any) => void
  onDragOver?: (e: any) => void
  onDragEnter?: (e: any) => void
  onDragLeave?: (e: any) => void
  onDragEnd?: (e: any) => void
  onDrop?: (e: any) => void
  style?: any
}

const flexVariations: { [index: string]: any } = {
  auto: css``,
  centralized: css`
    justify-content: center;
    align-items: center;
  `,
  fullCentralized: css`
    justify-content: center;
    align-items: center;
    height: 100vh;
  `
}

const FlexStyled = styled.div<Props>`
  ${space}
  ${fontSize}
  ${fontWeight}
  ${border}
  ${position}
  ${color}
  ${flexbox}
  ${flex}

  display: flex;
  
  ${({ variant }) => flexVariations[variant || 'auto']}

  ${({ fullHeight }) =>
    !!fullHeight &&
    css`
      height: 100vh;
    `}

  ${layout}
`

export const Flex = React.forwardRef<HTMLInputElement, Props>(
  ({ children, color, ...props }, ref) => {
    return (
      <FlexStyled ref={ref} color={color} {...props}>
        {children}
      </FlexStyled>
    )
  }
)

Flex.displayName = 'Flex'

Flex.propTypes = {
  ref: PropTypes.func,
  color: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node
}
