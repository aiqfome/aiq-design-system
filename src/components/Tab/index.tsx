export * from './Tabs'
export * from './Tab'
export * from './TabPanel'
