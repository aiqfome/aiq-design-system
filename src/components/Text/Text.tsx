import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  layout,
  LayoutProps,
  fontSize,
  FontSizeProps,
  fontWeight,
  FontWeightProps,
  typography,
  TypographyProps
} from 'styled-system'

export interface Props
  extends ColorProps,
    SpaceProps,
    LayoutProps,
    FontSizeProps,
    FontWeightProps,
    TypographyProps {
  cursor?: string
  whiteSpace?: 'nowrap' | 'normal' | 'pre'
}

export const Text = styled.span<Props>`
  ${color}
  ${space}
  ${layout}
  ${fontSize}
  ${fontWeight}
  ${typography}

  ${({ cursor }) =>
    cursor &&
    css`
      cursor: ${cursor};
    `};
  white-space: ${props => props.whiteSpace || 'normal'};
`

Text.propTypes = {
  cursor: PropTypes.string,
  whiteSpace: PropTypes.oneOf(['nowrap', 'normal', 'pre'])
}
