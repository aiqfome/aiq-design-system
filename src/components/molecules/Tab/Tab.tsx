import React from 'react'
import PropTypes from 'prop-types'
import { space, SpaceProps } from 'styled-system'
import styled, { css } from 'styled-components'

export interface TabProps extends SpaceProps {
  index: number
  children: any
  value?: number
  active?: boolean
  variant?: 'default' | 'contained' | 'card'
}

interface StyledProps {
  active?: boolean
  variant?: 'default' | 'contained' | 'card'
}

const tabVariations: { [index: string]: any } = {
  default: css<StyledProps>`
    padding: 8px 21px;
    ${({ active, theme }) =>
      active
        ? css`
            font-weight: ${theme.fontWeights.medium};
            color: ${theme.colors.primary};
            &::before {
              background: ${theme.colors.primary};
              animation: show 0.25s;
              @keyframes show {
                from {
                  transform: scale(0);
                }
                to {
                  transform: scale(1);
                }
              }
            }
          `
        : css`
            font-weight: ${theme.fontWeights.regular};
            color: ${theme.colors.darkGrey};
            &::before {
              background: transparent;
            }
          `}
  `,
  contained: css<StyledProps>`
    padding: 10px 17px;
    ${({ active, theme }) =>
      active
        ? css`
            font-weight: ${theme.fontWeights.semiBold};
            color: ${theme.colors.black};
            background: ${theme.colors.white};
            border-radius: 5px;
            border: 1px solid ${theme.colors.mediumGrey};
          `
        : css`
            font-weight: ${theme.fontWeights.regular};
            color: ${theme.colors.darkGrey};
          `}
  `,
  card: css<StyledProps>`
    padding: 10px 17px;
    ${({ active, theme }) =>
      active
        ? css`
            color: ${theme.colors.almostBlack};
            background: ${theme.colors.white};
            border-radius: 5px 5px 0 0;
            border: 1px solid ${theme.colors.mediumGrey};
            border-bottom: 0px;
            color: ${theme.colors.primary};
            font-weight: ${theme.fontWeights.medium};

            &::before {
              background: ${theme.colors.white};
            }
          `
        : css`
            font-weight: ${theme.fontWeights.regular};
            color: ${theme.colors.darkGrey};
          `}
  `
}

const TabStyled = styled.li<StyledProps>`
  ${space}

  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;

  &:hover {
    cursor: pointer;
  }

  &::before {
    position: absolute;
    content: '';
    bottom: -1px;
    width: 100%;
    height: 3px;
    border-radius: 3px 3px 0 0;
  }

  ${({ variant }) => tabVariations[variant || 'default']}
`

export const Tab: React.FC<TabProps> = ({
  children,
  value = 0,
  index,
  ...props
}) => {
  function handleClick(event) {
    event.currentTarget.parentNode.setAttribute('data-id', index)
  }

  return (
    <TabStyled onClick={handleClick} active={index === value} {...props}>
      {children}
    </TabStyled>
  )
}

Tab.propTypes = {
  index: PropTypes.number.isRequired,
  children: PropTypes.any.isRequired,
  value: PropTypes.number,
  variant: PropTypes.oneOf(['default', 'contained', 'card'])
}

Tab.defaultProps = {
  variant: 'default'
}
