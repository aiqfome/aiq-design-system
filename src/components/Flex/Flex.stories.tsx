import React from 'react'
import { withKnobs, select, text } from '@storybook/addon-knobs'

import { Flex } from './Flex'
import { Text } from '../Text'

export default {
  component: Flex,
  title: 'Flex',
  decorators: [withKnobs]
}

export const basic: React.FC = () => (
  <Flex
    justifyContent={select(
      'justifyContent',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    alignItems={select(
      'alignItems',
      { FlexStart: 'flex-start', FlexEnd: 'flex-end', Center: 'center' },
      'flex-start'
    )}
    flexDirection={select(
      'flexDirection',
      { Row: 'row', Column: 'column' },
      'row'
    )}
    backgroundColor={text('backgroundColor', '#fff')}
    variant={select(
      'Variant',
      { Centralized: 'centralized', FullCentralized: 'fullCentralized' },
      'centralized'
    )}
  >
    <Text>Flex!</Text>
  </Flex>
)

export const centralized: React.FC = () => (
  <Flex variant='centralized'>
    <Text>Flex!</Text>
  </Flex>
)

export const fullCentralized: React.FC = () => (
  <Flex variant='fullCentralized'>
    <Text>Flex!</Text>
  </Flex>
)
