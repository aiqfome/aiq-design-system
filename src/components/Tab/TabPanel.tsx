import React from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'

import { Flex, Props as FlexProps } from '../Flex'

export interface TabPanelProps extends FlexProps {
  value: number
  index: number
  children?: any
}

const TabPanelStyled = styled(Flex)`
  display: block;
`

export const TabPanel: React.FC<TabPanelProps> = ({
  children,
  value,
  index,
  ...props
}) => {
  return value === index ? (
    <TabPanelStyled {...props}>{children}</TabPanelStyled>
  ) : null
}

TabPanel.propTypes = {
  value: PropTypes.number.isRequired,
  index: PropTypes.number.isRequired,
  children: PropTypes.any
}
