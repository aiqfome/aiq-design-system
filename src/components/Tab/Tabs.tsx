import React, { useEffect, useState, useRef } from 'react'
import PropTypes from 'prop-types'
import { SpaceProps, space, LayoutProps, layout } from 'styled-system'
import styled, { css, DefaultTheme } from 'styled-components'

import { Flex } from '../Flex'

export interface TabsProps extends SpaceProps, LayoutProps, DefaultTheme {
  extra?: any
  children?: any
  wrapperProps?: any
  isMobile?: boolean
  variant?: 'default' | 'contained' | 'card'
  scrollPosition?: 'left' | 'middle' | 'right'
  onChange?: (event: any, newValue: any) => void
}

const tabsVariations: { [index: string]: any } = {
  default: css`
    &::before {
      position: absolute;
      content: '';
      left: 0px;
      bottom: 0px;
      width: 100%;
      height: 1px;
      background: #d9d9d9;
    }
  `,
  contained: css`
    background: #f5f5f5;
    border-radius: 5px;
    padding: 4px 5px;
  `,
  card: css`
    margin-bottom: 0;
  `
}

const TabStyled = styled.ul<TabsProps>`
  display: flex;
  flex-direction: row;
  list-style: none;
  margin-bottom: 20px;
  align-items: center;
  width: 100%;
  position: relative;

  ${space}
  ${layout}

  ${({ variant }) => tabsVariations[variant || 'default']}
`

const FlexStyled = styled(Flex)<TabsProps>`
  position: relative;
  display: flex;
  flex: 1;
  overflow: hidden;

  ::before {
    position: absolute;
    z-index: 1;
    opacity: 0;
    -webkit-transition: opacity 0.3s;
    transition: opacity 0.3s;
    content: '';
    pointer-events: none;
    left: 0;
    -webkit-box-shadow: inset 10px 0 8px -8px rgba(0, 0, 0, 0.39);
    box-shadow: inset 10px 0 8px -8px rgba(0, 0, 0, 0.39);
    top: 0;
    bottom: 0;
    width: 30px;
  }

  ::after {
    position: absolute;
    z-index: 1;
    opacity: 0;
    -webkit-transition: opacity 0.3s;
    transition: opacity 0.3s;
    content: '';
    pointer-events: none;
    right: 0;
    -webkit-box-shadow: inset -10px 0 8px -8px rgba(0, 0, 0, 0.39);
    box-shadow: inset -10px 0 8px -8px rgba(0, 0, 0, 0.39);
    top: 0;
    bottom: 0;
    width: 30px;
  }

  ${({ scrollPosition, isMobile }) => {
    if (isMobile) {
      if (scrollPosition === 'left') {
        return css`
          ::after {
            opacity: 1;
          }
        `
      }

      if (scrollPosition === 'right') {
        return css`
          ::before {
            opacity: 1;
          }
        `
      }

      return css`
        ::after {
          opacity: 1;
        }

        ::before {
          opacity: 1;
        }
      `
    }
  }}
`

export const Tabs: React.FC<TabsProps> = ({
  extra,
  children,
  wrapperProps,
  onChange = () => {
    // do nothing.
  },
  ...props
}) => {
  const [isMobile, setIsMobile] = useState(false)
  const [scrollPosition, setScrollPosition] = useState<
    'left' | 'middle' | 'right'
  >('left')

  const refFlex = useRef(document.createElement('div'))

  useEffect(() => {
    const handleResize = () => {
      if (
        refFlex.current &&
        refFlex.current.scrollWidth !== refFlex.current.clientWidth
      ) {
        setIsMobile(true)
      } else {
        setIsMobile(false)
      }
    }

    if (refFlex) {
      handleResize()
      window.addEventListener('resize', handleResize)
    }
  }, [refFlex])

  useEffect(() => {
    const handleScroll = e => {
      if (refFlex && isMobile) {
        refFlex.current.scrollLeft += e.deltaY

        if (refFlex.current.scrollLeft === 0) {
          setScrollPosition('left')
        } else if (
          refFlex.current.offsetWidth + refFlex.current.scrollLeft ===
          refFlex.current.scrollWidth
        ) {
          setScrollPosition('right')
        } else {
          setScrollPosition('middle')
        }
      }
    }

    if (refFlex) {
      refFlex.current.addEventListener('wheel', handleScroll)
    }
  }, [refFlex, isMobile])

  function handleClick(event) {
    onChange(event, parseInt(event.currentTarget.dataset.id))
  }

  return (
    <TabStyled {...props}>
      <FlexStyled
        isMobile={isMobile}
        scrollPosition={scrollPosition}
        {...wrapperProps}
      >
        <Flex
          flex={1}
          overflow='hidden'
          position='relative'
          onClick={handleClick}
          ref={refFlex}
        >
          {children}
        </Flex>
      </FlexStyled>

      {extra && <Flex px='12px'>{extra}</Flex>}
    </TabStyled>
  )
}

Tabs.propTypes = {
  extra: PropTypes.any,
  children: PropTypes.any,
  onChange: PropTypes.func,
  wrapperProps: PropTypes.object,
  variant: PropTypes.oneOf(['default', 'contained', 'card'])
}

Tabs.defaultProps = {
  variant: 'default'
}
