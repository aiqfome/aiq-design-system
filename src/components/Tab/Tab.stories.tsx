import React, { useState } from 'react'

import { Text } from '../Text'
import { Flex } from '../Flex'
import { Tabs } from './Tabs'
import { Tab } from './Tab'
import { TabPanel } from './TabPanel'

export default {
  component: Tab,
  title: 'Tab'
}

export const Basic: React.FC = () => {
  const [value, setValue] = useState(0)

  const handleChange = (event, newValue) => {
    setValue(newValue)
  }

  return (
    <Flex flexDirection='column'>
      <Tabs value={value} onChange={handleChange} width='100%'>
        <Tab value={value} index={0}>
          <Text fontSize={5} cursor='pointer'>
            Item One
          </Text>
        </Tab>
        <Tab value={value} index={1}>
          <Text fontSize={5} cursor='pointer'>
            Item Two
          </Text>
        </Tab>
        <Tab value={value} index={2}>
          <Text fontSize={5} cursor='pointer'>
            Item Three
          </Text>
        </Tab>
      </Tabs>
      <TabPanel value={value} index={0}>
        Item One
      </TabPanel>
      <TabPanel value={value} index={1}>
        Item Two
      </TabPanel>
      <TabPanel value={value} index={2}>
        Item Three
      </TabPanel>
    </Flex>
  )
}

export const Contained: React.FC = () => {
  const [value, setValue] = useState(0)

  const handleChange = (event, newValue) => {
    setValue(newValue)
  }

  return (
    <Flex flexDirection='column'>
      <Tabs variant='contained' value={value} onChange={handleChange}>
        <Tab variant='contained' value={value} index={0}>
          <Text fontSize={2} cursor='pointer'>
            Item One
          </Text>
        </Tab>
        <Tab variant='contained' value={value} index={1}>
          <Text fontSize={2} cursor='pointer'>
            Item Two
          </Text>
        </Tab>
        <Tab variant='contained' value={value} index={2}>
          <Text fontSize={2} cursor='pointer'>
            Item Three
          </Text>
        </Tab>
      </Tabs>
      <TabPanel value={value} index={0}>
        Item One
      </TabPanel>
      <TabPanel value={value} index={1}>
        Item Two
      </TabPanel>
      <TabPanel value={value} index={2}>
        Item Three
      </TabPanel>
    </Flex>
  )
}
