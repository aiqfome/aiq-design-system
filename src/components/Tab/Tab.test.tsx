import React, { useState } from 'react'

import { Text } from '../Text'
import { Flex } from '../Flex'
import { Tabs } from './Tabs'
import { Tab } from './Tab'
import { TabPanel } from './TabPanel'

import { render } from '../utils/test/render'

export const TabContent: React.FC = () => {
  const [value, setValue] = useState(0)

  const handleChange = (event, newValue) => {
    setValue(newValue)
  }

  return (
    <Flex flexDirection='column'>
      <Tabs onChange={handleChange}>
        <Tab value={value} index={0}>
          <Text fontSize={5} cursor='pointer'>
            Item One
          </Text>
        </Tab>
        <Tab value={value} index={1}>
          <Text fontSize={5} cursor='pointer'>
            Item Two
          </Text>
        </Tab>
        <Tab value={value} index={2}>
          <Text fontSize={5} cursor='pointer'>
            Item Three
          </Text>
        </Tab>
      </Tabs>
      <TabPanel value={value} index={0}>
        Item One
      </TabPanel>
      <TabPanel value={value} index={1}>
        Item Two
      </TabPanel>
      <TabPanel value={value} index={2}>
        Item Three
      </TabPanel>
    </Flex>
  )
}

describe('must match with the previous snapshot', () => {
  test('snapshot renders', () => {
    const component = render(<TabContent />)

    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
