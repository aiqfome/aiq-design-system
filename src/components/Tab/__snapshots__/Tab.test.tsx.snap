// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<div
  className="sc-gtssRu cBCJiK"
>
  <ul
    className="sc-dlnjPT cfWhEs"
  >
    <div
      className="sc-gtssRu cQyKxW sc-hKFyIo jYZvMs"
    >
      <div
        className="sc-gtssRu ldkLXJ"
        onClick={[Function]}
        overflow="hidden"
      >
        <li
          className="sc-eCApGN bWBJJc"
          onClick={[Function]}
        >
          <span
            className="sc-bdnylx jesgQY"
            cursor="pointer"
            fontSize={5}
          >
            Item One
          </span>
        </li>
        <li
          className="sc-eCApGN iFqYSf"
          onClick={[Function]}
        >
          <span
            className="sc-bdnylx jesgQY"
            cursor="pointer"
            fontSize={5}
          >
            Item Two
          </span>
        </li>
        <li
          className="sc-eCApGN iFqYSf"
          onClick={[Function]}
        >
          <span
            className="sc-bdnylx jesgQY"
            cursor="pointer"
            fontSize={5}
          >
            Item Three
          </span>
        </li>
      </div>
    </div>
  </ul>
  <div
    className="sc-gtssRu cQyKxW sc-jSFkmK ehsqFt"
  >
    Item One
  </div>
</div>
`;
